internals/overlord/state/notices_test.go

```diff
@@ -176,30 +219,72 @@ func (s *noticesSuite) testRepeatAfter(c *C, first, second, delay time.Duration)
 	c.Assert(newLastRepeated.After(lastRepeated), Equals, true)
 }
 
+func (s *noticesSuite) TestNoticesFilterUserID(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	uid1 := uint32(1000)
+	uid2 := uint32(0)
+	addNotice(c, st, &uid1, state.CustomNotice, "foo.com/bar", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, &uid2, state.ChangeUpdateNotice, "123", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, &uid2, state.WarningNotice, "Warning 1!", nil)
+	time.Sleep(time.Microsecond)
+	addNotice(c, st, nil, state.WarningNotice, "Warning 2!", nil)
+
+	// No user IDs
+	notices := st.Notices(nil)
+	c.Assert(notices, HasLen, 4)
+
+	// User ID nil
+	notices = st.Notices(&state.NoticeFilter{})
+	c.Assert(notices, HasLen, 4)
+
+	// One user ID
+	notices = st.Notices(&state.NoticeFilter{UserID: &uid2})
+	c.Assert(notices, HasLen, 3)
+	n := noticeToMap(c, notices[0])
+	c.Check(n["user-id"], Equals, float64(uid2))
+	c.Check(n["type"], Equals, "change-update")
+	c.Check(n["key"], Equals, "123")
+	n = noticeToMap(c, notices[1])
+	c.Check(n["user-id"], Equals, float64(uid2))
+	c.Check(n["type"], Equals, "warning")
+	c.Check(n["key"], Equals, "Warning 1!")
+	n = noticeToMap(c, notices[2])
+	c.Check(n["user-id"], Equals, nil)
+	c.Check(n["type"], Equals, "warning")
+	c.Check(n["key"], Equals, "Warning 2!")
+}
+
 func (s *noticesSuite) TestNoticesFilterType(c *C) {
 	st := state.New(nil)
 	st.Lock()
 	defer st.Unlock()
 
-	addNotice(c, st, state.CustomNotice, "foo.com/bar", nil)
+	addNotice(c, st, nil, state.CustomNotice, "foo.com/bar", nil)
 	time.Sleep(time.Microsecond)
-	addNotice(c, st, state.ChangeUpdateNotice, "123", nil)
+	addNotice(c, st, nil, state.ChangeUpdateNotice, "123", nil)
 	time.Sleep(time.Microsecond)
-	addNotice(c, st, state.WarningNotice, "Warning 1!", nil)
+	addNotice(c, st, nil, state.WarningNotice, "Warning 1!", nil)
 	time.Sleep(time.Microsecond)
-	addNotice(c, st, state.WarningNotice, "Warning 2!", nil)
+	addNotice(c, st, nil, state.WarningNotice, "Warning 2!", nil)
 
 	// No types
-	notices := st.Notices(nil)
+	notices := st.Notices(&state.NoticeFilter{})
```

Comment:
Out of interest, why this change? `nil` for the filter arg should still work, right? I think we should allow this (and test for it). Similar below.

Reply:
I wanted to ensure that a non-`nil` `NoticeFilter` with a `nil` `UserID` would correctly allow notices with any user ID. A `nil` `NoticeFilter` is a bit of a short circuit, allowing anything. I noticed that all the other filters also just checked with `nil` `NoticeFilter`, rather than `nil` `Types` etc. in a non-`nil` filter.

What I actually want is for each filter type, to test with a `nil` filter, a non-`nil` filter with all `nil` fields, and then the set filters we care about testing. I'll tweak this in the next commit.

Reply:
Thanks -- fine with me then.